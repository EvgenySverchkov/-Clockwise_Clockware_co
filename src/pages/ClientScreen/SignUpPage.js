import React from "react";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";

import { changeSignUpIsLoad, changeSuccessModalData, changeWarningModalData } from "../../store/clientSide/services/actions";

import { SERVERDOMAIN } from "../../services/serverUrls";

import FormGroup from "../../components/FormComponents/FormGroup";
import Label from "../../components/FormComponents/Label";
import TextField from "../../components/FormComponents/TextField";
import EmailField from "../../components/FormComponents/EmailField";
import PasswordField from "../../components/FormComponents/PasswordField";
import Button from "../../components/FormComponents/Button";

function RegistrationForm({history}) {
  const state = useSelector(state=>{
    return {signUpIsLoad: state.client_services.signUpIsLoad}
  });
  const dispatch = useDispatch();

  function handler(e) {
    e.preventDefault();
    const elem = e.target;
    let newObj = {
      name: elem.name.value,
      email: elem.email.value,
      password: elem.password.value,
    };
    dispatch(changeSignUpIsLoad(true));
    fetch(`${SERVERDOMAIN}/signUp`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json;charset=utf-8",
      },
      body: JSON.stringify(newObj),
    })
    .then((data) => data.json())
    .then((data) => {
      dispatch(changeSignUpIsLoad(false));
        if (!data.success) {
          dispatch(changeWarningModalData({
            msg: data.msg
          }));
          document.getElementById("callWarningModalBtn").click();
        } else {
          dispatch(changeSuccessModalData({
            msg: `Congratulations! ${data.user.name} you are registered`, 
            backBtnTxt: "Back to the main page", 
            backTo: "/"
          }));
          document.getElementById("callSuccessModalBtn").click();
          history.push("/client/login");
        }
    });
  }
  return (
    <form onSubmit={handler} className="mt-4 row justify-content-center">
      <FormGroup isRow={true}>
        <Label forId={"name"}>Enter your name</Label>
        <TextField id={"name"} name={"name"} />
      </FormGroup>
      <FormGroup isRow={true}>
        <Label forId={"email"}>Enter your email</Label>
        <EmailField id={"email"} />
      </FormGroup>
      <FormGroup isRow={true}>
        <Label forId={"password"}>Enter your password</Label>
        <PasswordField id={"password"} />
      </FormGroup>
      <Button isLoad={state.signUpIsLoad} value={"Sign Up"} />
    </form>
  );
}

RegistrationForm.propTypes = {
  history: PropTypes.object.isRequired,
};

export default RegistrationForm;
