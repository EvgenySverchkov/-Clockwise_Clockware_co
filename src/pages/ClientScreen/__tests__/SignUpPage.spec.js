import React from "react";
import {create, act} from "react-test-renderer";
import {Provider} from "react-redux";
import configMockStore from "redux-mock-store";

import Context from "../../../ContextComponent";
import SignUpPage from "../SignUpPage";

const mockStore = configMockStore();


describe("Test of <SignUpPage/>", ()=>{
    it("Default view", ()=>{
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const store = mockStore({})
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <SignUpPage />
                </Context.Provider>
            </Provider>
        )
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("Call submit event when one field of form is empty", ()=>{
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const store = mockStore({})
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                name: {
                    value: ""
                },
                email: {
                    value: "validMail@mail.com"
                },
                password: {
                    value: "12345678"
                }
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <SignUpPage />
                </Context.Provider>
            </Provider>
        )
        const handlerReturn = component.root.findByType("form").props.onSubmit(eventObject);
        expect(mockContextValue.openErrorWindowWithMsg.mock.calls.length).toBe(1);
        expect(handlerReturn).toBeFalsy();
    });
    it("Call submit event when name field is invalid", ()=>{
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const store = mockStore({})
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                name: {
                    value: "Invalid1234"
                },
                email: {
                    value: "validMail@mail.com"
                },
                password: {
                    value: "12345678"
                }
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <SignUpPage />
                </Context.Provider>
            </Provider>
        )
        const handlerReturn = component.root.findByType("form").props.onSubmit(eventObject);
        expect(mockContextValue.openWarningTooltip.mock.calls.length).toBe(1);
        expect(handlerReturn).toBeFalsy();
    });
    it("Call submit event when email field is invalid", ()=>{
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const store = mockStore({})
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                name: {
                    value: "ValidName"
                },
                email: {
                    value: "invalidMailmail.com"
                },
                password: {
                    value: "12345678"
                }
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <SignUpPage />
                </Context.Provider>
            </Provider>
        )
        const handlerReturn = component.root.findByType("form").props.onSubmit(eventObject);
        expect(mockContextValue.openWarningTooltip.mock.calls.length).toBe(1);
        expect(handlerReturn).toBeFalsy();
    });
    it("Call submit event when email field containt word 'admin'", ()=>{
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const store = mockStore({})
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                name: {
                    value: "ValidName"
                },
                email: {
                    value: "admin@mail.com"
                },
                password: {
                    value: "12345678"
                }
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <SignUpPage />
                </Context.Provider>
            </Provider>
        )
        const handlerReturn = component.root.findByType("form").props.onSubmit(eventObject);
        expect(mockContextValue.openWarningTooltip.mock.calls.length).toBe(1);
        expect(handlerReturn).toBeFalsy();
    });
    it("Call submit event when password field is invalid", ()=>{
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const store = mockStore({})
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                name: {
                    value: "ValidName"
                },
                email: {
                    value: "valid@mail.com"
                },
                password: {
                    value: "1"
                }
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <SignUpPage />
                </Context.Provider>
            </Provider>
        )
        const handlerReturn = component.root.findByType("form").props.onSubmit(eventObject);
        expect(mockContextValue.openWarningTooltip.mock.calls.length).toBe(1);
        expect(handlerReturn).toBeFalsy();
    });
    it("Call submit event when all field are valid and server return object with success fild with value true and with user data", async ()=>{
        global.fetch = jest.fn(()=>Promise.resolve({
            json: jest.fn(()=>Promise.resolve({
                success: true,
                user: {
                    name: 'Name'
                }
            }))
        }));
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const store = mockStore({})
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                name: {
                    value: "Name"
                },
                email: {
                    value: "valid@mail.com"
                },
                password: {
                    value: "1234567"
                },
                reset: jest.fn()
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <SignUpPage />
                </Context.Provider>
            </Provider>
        )
        await act(async ()=>{
            await component.root.findByType("form").props.onSubmit(eventObject);
        })
        expect(mockContextValue.openSuccessWindowWithMsg.mock.calls.length).toBe(1);
        expect(eventObject.target.reset.mock.calls.length).toBe(1)
    });
    it("Call submit event when all field are valid but server return object with success fild with value false", async ()=>{
        global.fetch = jest.fn(()=>Promise.resolve({
            json: jest.fn(()=>Promise.resolve({
                success: false
            }))
        }));
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const store = mockStore({})
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                name: {
                    value: "Name"
                },
                email: {
                    value: "valid@mail.com"
                },
                password: {
                    value: "1234567"
                },
                reset: jest.fn()
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <SignUpPage />
                </Context.Provider>
            </Provider>
        )
        await act(async ()=>{
            await component.root.findByType("form").props.onSubmit(eventObject);
        })
        expect(mockContextValue.openErrorWindowWithMsg.mock.calls.length).toBe(1);
    });
});