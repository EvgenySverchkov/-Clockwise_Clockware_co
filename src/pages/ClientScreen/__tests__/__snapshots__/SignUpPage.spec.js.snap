// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <SignUpPage/> Default view 1`] = `
<form
  className="mt-4 row justify-content-center"
  onBlur={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="name"
    >
      Enter your name
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Warning"
        type="text"
      />
    </div>
  </div>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="email"
    >
      Enter your email
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        readOnly={false}
        required={true}
        title="Warning"
        type="email"
      />
    </div>
  </div>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="password"
    >
      Enter your password
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        autoComplete="username"
        className="form-control"
        id="password"
        name="password"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Warning"
        type="password"
      />
    </div>
  </div>
  <div
    className="row justify-content-sm-center col-12"
  >
    <input
      className="btn btn-primary col-12 col-sm-4 mt-3"
      type="submit"
      value="Sign Up"
    />
  </div>
</form>
`;
